File: C/src/bin/easy_47.c
// Daily Coding Problem: 47
// Difficulty: Easy
// Asked by: Amazon
// Authored by: TenthEdict

#include <stdio.h>

int best_profit(int *prices, int length) {
    if (length < 2) {
        return 0;
    }

    int profit = 0;
    int min_price = prices[0];

    for (int i = 1; i < length; i++) {
        if (prices[i] < min_price) {
            min_price = prices[i];
        }
        profit = (profit > prices[i] - min_price) ? profit : prices[i] - min_price;
    }
    return profit;
}

int main() {
    int case1[6] = {9, 11, 8, 5, 7, 10};
    printf("%d\n", best_profit(case1, 6));
    return 0;
}

File: Rust/src/bin/easy_47.rs
// Daily Coding Problem: 47
// Difficulty: Easy
// Asked by: Amazon
// Author: TenthEdict

fn best_profit(arr: &[i32]) -> Option<i32> {
    if arr.len() < 2 {
        return Some(0);
    }

    let mut min_price: i32 = arr[0];
    let mut profit: i32 = 0;

    for price in arr {
        if *price < min_price {
            min_price = *price;
        } 
        profit = profit.max(*price - min_price);
    }
    Some(profit)
}

fn main() {
    let cases: Vec<(&[i32], i32)> = vec![
        (&[9, 11, 8, 5, 7, 10], 5),
        (&[10, 2, 11, 1, 5], 9),
        (&[10, 8, 5, 2], 0),
        (&[5], 0),
        (&[], 0),
    ];

    for (arr, expected) in cases {
        let result = best_profit(arr);
        let pass = if result == Some(expected) { "✓" } else { "✗" };
        println!("{:?} -> {:?} (expected {}) {}", arr, result, expected, pass);
    }
}
